# Compare the performances between base model and LoRA model
Qwen3-4B-Instruct with Flash-attention

## 1. Preparation

### 1.1 Import libraries

In [1]:
import os
import gc
from datetime import datetime
import json
import ast
import re
from tqdm import tqdm
import torch
from datasets import load_dataset, DatasetDict
from peft import PeftModel

from transformers import (
    AutoTokenizer,
    AutoModelForCausalLM,
    DataCollatorForSeq2Seq,
    TrainingArguments,
    Trainer,
    set_seed
)

from peft import (
    LoraConfig,
    get_peft_model,
    TaskType,
    prepare_model_for_kbit_training
)

from transformers.utils.notebook import NotebookProgressCallback

In [2]:
import warnings
warnings.filterwarnings('ignore')

### 1.2 Set constants

In [3]:
MODEL_NAME = "../models/Qwen/Qwen3-4B-Instruct-2507"
DATA_PATH = "../data/final/LoRA_samples_sum_fixed.jsonl"

MAX_LENGTH = 2048
SEED = 42

set_seed(SEED)

### 1.3 Create dataset

Load and split the dataset with exactly the same approach as the training. This is to make sure we are using the same test set without data leakage.

In [4]:
ds_raw = load_dataset("json", data_files=DATA_PATH)["train"]
shuffled_ds = ds_raw.shuffle(seed=SEED)

In [5]:
train_temp = shuffled_ds.train_test_split(test_size=0.25, seed=SEED)

val_test = train_temp['test'].train_test_split(test_size=0.4, seed=42)

ds = DatasetDict({
    'train': train_temp['train'], # 75%
    'validation': val_test['train'], # 15%
    'test': val_test['test'] # 10%
})

Set up tokenizer and tokenized datasets with exactly the same approach as the training.

In [6]:
tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME, trust_remote_code=True)

In [7]:
tokenizer.padding_side = "right"
tokenizer.truncation_side = "left"

Set up the process function for correct tokenization for the data samples.

In [8]:
def process_func(example):

    messages = example["messages"]

    instruction_text = tokenizer.apply_chat_template(
        messages[:-1],
        tokenize=False,
        add_generation_prompt=True
    )

    full_text = tokenizer.apply_chat_template(
        messages,
        tokenize=False,
        add_generation_prompt=False
    ) + tokenizer.eos_token

    tokenized_full = tokenizer(
        full_text,
        max_length=MAX_LENGTH,
        truncation=True,
        add_special_tokens=False
    )

    tokenized_instruction = tokenizer(
        instruction_text,
        max_length=MAX_LENGTH,
        truncation=True,
        add_special_tokens=False
    )

    input_ids = tokenized_full["input_ids"]
    attention_mask = tokenized_full["attention_mask"]

    instruction_len = len(tokenized_instruction["input_ids"])

    labels = [-100] * instruction_len + input_ids[instruction_len:]
    labels = labels[:len(input_ids)]

    return {
        "input_ids": input_ids,
        "attention_mask": attention_mask,
        "labels": labels
    }

In [9]:
tokenized_ds = ds.map(
    process_func,
    remove_columns=ds["train"].column_names,
    num_proc=os.cpu_count(),
    desc="Processing dataset"
)

In [10]:
data_collator = DataCollatorForSeq2Seq(
    tokenizer=tokenizer,
    padding=True,
    pad_to_multiple_of=8,
    label_pad_token_id=-100
)

### 1.4 Helper functions

In [11]:
def extract_intention_id(response_text):
    """Extract intention_id with flexible JSON parsing"""
    response_text = response_text.strip()
    
    # Try 1: Standard JSON (double quotes)
    try:
        data = json.loads(response_text)
        if isinstance(data, dict):
            return data.get("intention_id", None)
    except:
        pass
    
    # Try 2: Python dict literal (handles single quotes)
    try:
        data = ast.literal_eval(response_text)
        if isinstance(data, dict):
            return data.get("intention_id", None)
    except:
        pass
    
    # Try 3: Regex fallback
    match = re.search(r'"intention_id"\s*:\s*"([^"]+)"', response_text)
    if not match:
        match = re.search(r"'intention_id'\s*:\s*'([^']+)'", response_text)
    if match:
        return match.group(1)
    
    return None


def parse_ground_truth(content):
    """Parse ground truth JSON with flexible quote handling"""
    content = content.strip()
    
    # Try 1: Standard JSON
    try:
        return json.loads(content)
    except:
        pass
    
    # Try 2: Python dict literal (single quotes)
    try:
        return ast.literal_eval(content)
    except:
        pass
    
    return None

In [12]:
def evaluate_intention_accuracy(model, tokenizer, test_data, name, batch_size=2):
    import json, ast, re, torch
    from tqdm import tqdm
    
    print(f"\nEvaluating {name} on intention_id...")
    
    # ✅ CRITICAL: Check test_data upfront
    print(f"  test_data type: {type(test_data)}")
    print(f"  test_data length: {len(test_data)}")
    if len(test_data) == 0:
        print("⚠️ ERROR: test_data is empty!")
        return {"accuracy": 0, "correct": 0, "total": 0, "skipped": 0}
    
    model.eval()
    
    # Save and set padding
    original_padding_side = tokenizer.padding_side
    tokenizer.padding_side = "left"
    
    if tokenizer.pad_token_id is None:
        tokenizer.pad_token_id = tokenizer.eos_token_id
    
    correct = 0
    total = 0
    invalid_responses = 0
    samples_skipped = 0
    
    # Track skip reasons
    skip_reasons = {"no_messages": 0, "json_parse": 0, "no_id": 0, "prompt_gen": 0, "other": 0}
    
    for i in tqdm(range(0, len(test_data), batch_size), desc="Evaluating"):
        batch_start = i
        batch_end = min(i + batch_size, len(test_data))
        
        prompts = []
        ground_truth_ids = []
        
        for idx in range(batch_start, batch_end):
            try:
                example = test_data[idx]
                messages = example.get("messages", [])
                
                # Check 1: Messages structure
                if not messages or len(messages) < 2:
                    skip_reasons["no_messages"] += 1
                    samples_skipped += 1
                    continue
                
                # Check 2: Parse ground truth
                last_msg_content = messages[-1].get("content", "").strip()
                
                ground_truth_json = None
                try:
                    ground_truth_json = json.loads(last_msg_content)
                except:
                    try:
                        ground_truth_json = ast.literal_eval(last_msg_content)
                    except Exception as e:
                        skip_reasons["json_parse"] += 1
                        samples_skipped += 1
                        # Print first 3 JSON errors
                        if samples_skipped <= 3:
                            print(f"⚠️ JSON error (sample {idx}): {e}")
                        continue
                
                gt_id = ground_truth_json.get("intention_id") if ground_truth_json else None
                if gt_id is None:
                    skip_reasons["no_id"] += 1
                    samples_skipped += 1
                    continue
                
                # Check 3: Generate prompt
                try:
                    prompt = tokenizer.apply_chat_template(
                        messages[:-1],
                        tokenize=False,
                        add_generation_prompt=True
                    )
                except Exception as e:
                    skip_reasons["prompt_gen"] += 1
                    samples_skipped += 1
                    if samples_skipped <= 3:
                        print(f"⚠️ Prompt error (sample {idx}): {e}")
                    continue
                
                prompts.append(prompt)
                ground_truth_ids.append(str(gt_id))
                
            except Exception as e:
                skip_reasons["other"] += 1
                samples_skipped += 1
                # ✅ CRITICAL: Print first 5 unexpected errors
                if samples_skipped <= 5:
                    print(f"⚠️ Unexpected error (sample {idx}): {type(e).__name__}: {e}")
                continue
        
        # Skip batch if no valid prompts
        if len(prompts) == 0:
            continue
        
        # Tokenize
        inputs = tokenizer(
            prompts,
            return_tensors="pt",
            padding=True,
            truncation=True,
            max_length=2048
        ).to(model.device)
        
        # Generate
        with torch.no_grad():
            outputs = model.generate(
                **inputs,
                max_new_tokens=128,
                do_sample=False,
                pad_token_id=tokenizer.pad_token_id,
                eos_token_id=tokenizer.eos_token_id
            )
        
        # Evaluate predictions
        for j, output in enumerate(outputs):
            prediction_text = tokenizer.decode(
                output[inputs.input_ids.shape[1]:],
                skip_special_tokens=True
            ).strip()
            
            predicted_id = extract_intention_id(prediction_text)
            
            if predicted_id is None:
                invalid_responses += 1
            elif str(predicted_id) == str(ground_truth_ids[j]):
                correct += 1
            total += 1
            
            # Print first 3 predictions for verification
            if total <= 100:
                print(f"\n--- Prediction {total} ---")
                print(f"Expected: {ground_truth_ids[j]}")
                print(f"Predicted: {predicted_id}")
                print(f"Raw: {prediction_text[:100]}...")
    
    # Restore padding
    tokenizer.padding_side = original_padding_side
    
    # Print skip reasons
    print(f"\n🔍 Skip reasons: {skip_reasons}")
    
    accuracy = (correct / total) * 100 if total > 0 else 0
    valid_rate = ((total - invalid_responses) / total) * 100 if total > 0 else 0
    
    print(f"\n{'='*60}")
    print(f"{name} Results:")
    print(f"{'='*60}")
    print(f"  Total processed: {total}")
    print(f"  Samples skipped: {samples_skipped}")
    print(f"  Correct: {correct}/{total} ({accuracy:.2f}%)")
    print(f"  Valid responses: {valid_rate:.2f}%")
    print(f"{'='*60}")
    
    return {
        "accuracy": accuracy,
        "correct": correct,
        "total": total,
        "invalid": invalid_responses,
        "valid_rate": valid_rate,
        "skipped": samples_skipped,
        "skip_reasons": skip_reasons
    }

## 2. Test the base model

### 2.1 Load base model and configure the test

In [13]:
base_model = AutoModelForCausalLM.from_pretrained(
    MODEL_NAME,
    dtype=torch.bfloat16,
    device_map="auto",
    trust_remote_code=True,
    attn_implementation="flash_attention_2"
)

Loading weights:   0%|          | 0/398 [00:00<?, ?it/s]

In [14]:
base_model.eval()

Qwen3ForCausalLM(
  (model): Qwen3Model(
    (embed_tokens): Embedding(151936, 2560)
    (layers): ModuleList(
      (0-35): 36 x Qwen3DecoderLayer(
        (self_attn): Qwen3Attention(
          (q_proj): Linear(in_features=2560, out_features=4096, bias=False)
          (k_proj): Linear(in_features=2560, out_features=1024, bias=False)
          (v_proj): Linear(in_features=2560, out_features=1024, bias=False)
          (o_proj): Linear(in_features=4096, out_features=2560, bias=False)
          (q_norm): Qwen3RMSNorm((128,), eps=1e-06)
          (k_norm): Qwen3RMSNorm((128,), eps=1e-06)
        )
        (mlp): Qwen3MLP(
          (gate_proj): Linear(in_features=2560, out_features=9728, bias=False)
          (up_proj): Linear(in_features=2560, out_features=9728, bias=False)
          (down_proj): Linear(in_features=9728, out_features=2560, bias=False)
          (act_fn): SiLUActivation()
        )
        (input_layernorm): Qwen3RMSNorm((2560,), eps=1e-06)
        (post_attention_layer

In [15]:
eval_args = TrainingArguments(
    output_dir="./temp_eval",
    per_device_eval_batch_size=1,  # ⬇️ CRITICAL: Was likely 4 or 8
    dataloader_num_workers=0,       # ✅ Windows-safe
    bf16=True,                      # ✅ Match training precision
    report_to="none",
)

base_trainer = Trainer(
    model=base_model,
    args=eval_args,
    eval_dataset=tokenized_ds["test"],
    data_collator=data_collator
)

base_trainer.remove_callback(NotebookProgressCallback)

### 2.2 Perform the test

In [16]:
# Evaluate loss
print("Computing evaluation loss...")
base_metrics = base_trainer.evaluate()
base_loss = base_metrics["eval_loss"]
base_perplexity = torch.exp(torch.tensor(base_loss)).item()
print(f"Base Model - Loss: {base_loss:.4f}, Perplexity: {base_perplexity:.2f}")

Computing evaluation loss...
Base Model - Loss: 1.8264, Perplexity: 6.21


In [17]:
print("\nComputing generation accuracy...")

# ✅ Ensure left-padding for generation
original_padding = tokenizer.padding_side
tokenizer.padding_side = "left"

base_results = evaluate_intention_accuracy(
    base_model,
    tokenizer,
    ds["test"],  # Raw dataset
    "Base Qwen3-4B",
    batch_size=2  # ⬇️ Small batch for generation too
)

# Restore padding
tokenizer.padding_side = original_padding


Computing generation accuracy...

Evaluating Base Qwen3-4B on intention_id...
  test_data type: <class 'datasets.arrow_dataset.Dataset'>
  test_data length: 1011


Evaluating:   0%|          | 1/506 [00:03<28:05,  3.34s/it]


--- Prediction 1 ---
Expected: 1a2b3c4d5e6f7g8h
Predicted: 1a2b3c4d5e6f7g8h
Raw: {"input_summary": "用户不在佛山", "intention_id": "1a2b3c4d5e6f7g8h"}...

--- Prediction 2 ---
Expected: n0o1p2q3r4s5t6u7
Predicted: k7l8m9n0o1p2q3r4
Raw: {"input_summary": "用户发送不当言论", "intention_id": "k7l8m9n0o1p2q3r4"}...


Evaluating:   0%|          | 2/506 [00:05<22:36,  2.69s/it]


--- Prediction 3 ---
Expected: 9z8y7x6w5v4u3t2s
Predicted: 9z8y7x6w5v4u3t2s
Raw: {"input_summary": "用户询问收费", "intention_id": "9z8y7x6w5v4u3t2s"}...

--- Prediction 4 ---
Expected: 1a2b3c4d5e6f7g8h
Predicted: 1a2b3c4d5e6f7g8h
Raw: {"input_summary": "用户询问带猫入场", "intention_id": "1a2b3c4d5e6f7g8h"}...


Evaluating:   1%|          | 3/506 [00:07<20:25,  2.44s/it]


--- Prediction 5 ---
Expected: 1a2b3c4d5e6f7890
Predicted: 1a2b3c4d5e6f7890
Raw: {"input_summary": "用户拒绝参加活动", "intention_id": "1a2b3c4d5e6f7890"}...

--- Prediction 6 ---
Expected: a1b2c3d4e5f6g7h8
Predicted: a1b2c3d4e5f6g7h8
Raw: {"input_summary": "用户拒绝参加展会", "intention_id": "a1b2c3d4e5f6g7h8"}...


Evaluating:   1%|          | 4/506 [00:09<19:22,  2.32s/it]


--- Prediction 7 ---
Expected: a1b2c3d4e5f6g7h8
Predicted: k1j2h3g4f5d6s7a8
Raw: {"input_summary": "用户拒绝电话打扰", "intention_id": "k1j2h3g4f5d6s7a8"}...

--- Prediction 8 ---
Expected: h8i9j0k1l2m3n4o5
Predicted: h8i9j0k1l2m3n4o5
Raw: {"input_summary": "用户询问停车", "intention_id": "h8i9j0k1l2m3n4o5"}...


Evaluating:   1%|          | 5/506 [00:12<18:53,  2.26s/it]


--- Prediction 9 ---
Expected: d4e5f6g7h8i9j0k1l
Predicted: d4e5f6g7h8i9j0k1l
Raw: {"input_summary": "用户询问是否收费", "intention_id": "d4e5f6g7h8i9j0k1l"}...

--- Prediction 10 ---
Expected: l2_n3_o4_p5_q6_r7
Predicted: l2_n3_o4_p5_q6_r7
Raw: {"input_summary": "用户询问是否收费", "intention_id": "l2_n3_o4_p5_q6_r7"}...


Evaluating:   1%|          | 6/506 [00:14<18:41,  2.24s/it]


--- Prediction 11 ---
Expected: d8e7f6a5b4c3d2e1
Predicted: d8e7f6a5b4c3d2e1
Raw: {"input_summary": "用户询问预约方式", "intention_id": "d8e7f6a5b4c3d2e1"}...

--- Prediction 12 ---
Expected: others
Predicted: i9j0k1l2m3n4o5p6
Raw: {"input_summary": "用户骂人", "intention_id": "i9j0k1l2m3n4o5p6"}...


Evaluating:   1%|▏         | 7/506 [00:16<18:34,  2.23s/it]


--- Prediction 13 ---
Expected: y5z6a7b8c9d0e1f2
Predicted: y5z6a7b8c9d0e1f2
Raw: {"input_summary": "用户询问楼盘名单", "intention_id": "y5z6a7b8c9d0e1f2"}...

--- Prediction 14 ---
Expected: 9k0l1m2n3o4p5q6r
Predicted: 9k0l1m2n3o4p5q6r
Raw: {"input_summary": "用户肯定参加", "intention_id": "9k0l1m2n3o4p5q6r"}...


Evaluating:   2%|▏         | 8/506 [00:17<16:19,  1.97s/it]


--- Prediction 15 ---
Expected: others
Predicted: others
Raw: {"input_summary": "用户询问带宠物政策", "intention_id": "others"}...

--- Prediction 16 ---
Expected: c3d4e5f6g7h8i9j0
Predicted: others
Raw: {"input_summary": "用户询问报名方式", "intention_id": "others"}...


Evaluating:   2%|▏         | 9/506 [00:20<17:10,  2.07s/it]


--- Prediction 17 ---
Expected: 5e4f3a2b1c9d8e7f
Predicted: 5e4f3a2b1c9d8e7f
Raw: {"input_summary": "用户询问技术研讨会", "intention_id": "5e4f3a2b1c9d8e7f"}...

--- Prediction 18 ---
Expected: s9t0u1v2w3x4y5z6a
Predicted: s9t0u1v2w3x4y5z6a
Raw: {"input_summary": "用户询问媒体证", "intention_id": "s9t0u1v2w3x4y5z6a"}...


Evaluating:   2%|▏         | 10/506 [00:22<17:20,  2.10s/it]


--- Prediction 19 ---
Expected: b1c2d3e4f5g6h7i
Predicted: b1c2d3e4f5g6h7i
Raw: {"input_summary": "用户关心儿童区", "intention_id": "b1c2d3e4f5g6h7i"}...

--- Prediction 20 ---
Expected: others
Predicted: f90dcbadece18924
Raw: {"input_summary": "用户拒绝参展", "intention_id": "f90dcbadece18924"}...


Evaluating:   2%|▏         | 11/506 [00:24<17:49,  2.16s/it]


--- Prediction 21 ---
Expected: 5y6z7a8b9c0d1e2f
Predicted: 9g0h1i2j3k4l5m6n
Raw: {"input_summary": "用户询问地址", "intention_id": "9g0h1i2j3k4l5m6n"}...

--- Prediction 22 ---
Expected: k7007007007007007
Predicted: k7007007007007007
Raw: {"input_summary": "用户询问旧机回收", "intention_id": "k7007007007007007"}...


Evaluating:   2%|▏         | 12/506 [00:26<18:06,  2.20s/it]


--- Prediction 23 ---
Expected: kb_005_56789012
Predicted: b2c3d4e5f6g7h8i9
Raw: {"input_summary": "用户拒绝机器沟通", "intention_id": "b2c3d4e5f6g7h8i9"}...

--- Prediction 24 ---
Expected: others
Predicted: 9b8b23f618642a12
Raw: {"input_summary": "用户明确拒绝", "intention_id": "9b8b23f618642a12"}...


Evaluating:   3%|▎         | 13/506 [00:29<18:08,  2.21s/it]


--- Prediction 25 ---
Expected: b2c3d4e5f6g7h8i9
Predicted: b2c3d4e5f6g7h8i9
Raw: {"input_summary": "用户询问开始时间", "intention_id": "b2c3d4e5f6g7h8i9"}...

--- Prediction 26 ---
Expected: d7e6f5g4h3i2j1k0
Predicted: d7e6f5g4h3i2j1k0
Raw: {"input_summary": "用户询问表演", "intention_id": "d7e6f5g4h3i2j1k0"}...


Evaluating:   3%|▎         | 14/506 [00:30<15:59,  1.95s/it]


--- Prediction 27 ---
Expected: others
Predicted: others
Raw: {"input_summary": "用户询问酒店推荐", "intention_id": "others"}...

--- Prediction 28 ---
Expected: others
Predicted: others
Raw: {"input_summary": "用户询问现场餐饮", "intention_id": "others"}...


Evaluating:   3%|▎         | 15/506 [00:32<16:37,  2.03s/it]


--- Prediction 29 ---
Expected: q1w2e3r4t5y6u7i8
Predicted: q1w2e3r4t5y6u7i8
Raw: {"input_summary": "用户询问另一个品牌", "intention_id": "q1w2e3r4t5y6u7i8"}...

--- Prediction 30 ---
Expected: k005
Predicted: k001
Raw: {"input_summary": "用户要求获取信息", "intention_id": "k001"}...


Evaluating:   3%|▎         | 16/506 [00:34<16:45,  2.05s/it]


--- Prediction 31 ---
Expected: b3d8e1f4a6c9d7b2
Predicted: b3d8e1f4a6c9d7b2
Raw: {"input_summary": "用户需要时间考虑", "intention_id": "b3d8e1f4a6c9d7b2"}...

--- Prediction 32 ---
Expected: 9f8d7s6a5s4d3a2s
Predicted: 9f8d7s6a5s4d3a2s
Raw: {"input_summary": "用户拒绝参加", "intention_id": "9f8d7s6a5s4d3a2s"}...


Evaluating:   3%|▎         | 17/506 [00:36<16:47,  2.06s/it]


--- Prediction 33 ---
Expected: r8s7t6u5v4w3x2y1
Predicted: r8s7t6u5v4w3x2y1
Raw: {"input_summary": "用户询问门票", "intention_id": "r8s7t6u5v4w3x2y1"}...

--- Prediction 34 ---
Expected: others
Predicted: others
Raw: {"input_summary": "用户发送不当言论", "intention_id": "others"}...


Evaluating:   4%|▎         | 18/506 [00:39<17:11,  2.11s/it]


--- Prediction 35 ---
Expected: others
Predicted: f6g7h8i9j0k1l2m3n4
Raw: {"input_summary": "用户询问住宿", "intention_id": "f6g7h8i9j0k1l2m3n4"}...

--- Prediction 36 ---
Expected: m2n3b4v5c6x7z8l9k0
Predicted: p0o9i8u7y6t5r4e3w
Raw: {"input_summary": "用户询问收费", "intention_id": "p0o9i8u7y6t5r4e3w"}...


Evaluating:   4%|▍         | 19/506 [00:41<17:27,  2.15s/it]


--- Prediction 37 ---
Expected: a3b4c5d6e7f8a9b0
Predicted: a3b4c5d6e7f8a9b0
Raw: {"input_summary": "用户询问旧物回收", "intention_id": "a3b4c5d6e7f8a9b0"}...

--- Prediction 38 ---
Expected: 71e25d86c5f68440
Predicted: 71e25d86c5f68440
Raw: {"input_summary": "用户辱骂客服", "intention_id": "71e25d86c5f68440"}...


Evaluating:   4%|▍         | 20/506 [00:43<17:21,  2.14s/it]


--- Prediction 39 ---
Expected: 6f7g8h9i0j1k2l3m
Predicted: 6f7g8h9i0j1k2l3m
Raw: {"input_summary": "用户骂人", "intention_id": "6f7g8h9i0j1k2l3m"}...

--- Prediction 40 ---
Expected: others
Predicted: others
Raw: {"input_summary": "用户询问是否需身份证", "intention_id": "others"}...


Evaluating:   4%|▍         | 21/506 [00:45<17:22,  2.15s/it]


--- Prediction 41 ---
Expected: 3a2b1c9d8e7f6g5h
Predicted: 3a2b1c9d8e7f6g5h
Raw: {"input_summary": "用户询问买样品", "intention_id": "3a2b1c9d8e7f6g5h"}...

--- Prediction 42 ---
Expected: k_a1b2c3d4e5f6g7h
Predicted: k_a1b2c3d4e5f6g7h
Raw: {"input_summary": "用户询问门票费用", "intention_id": "k_a1b2c3d4e5f6g7h"}...


Evaluating:   4%|▍         | 22/506 [00:47<17:24,  2.16s/it]


--- Prediction 43 ---
Expected: h2j3k4l5m6n7b8v9c
Predicted: h2j3k4l5m6n7b8v9c
Raw: {"input_summary": "用户询问班车频率", "intention_id": "h2j3k4l5m6n7b8v9c"}...

--- Prediction 44 ---
Expected: a1b2c3d4e5f6g7h8
Predicted: a1b2c3d4e5f6g7h8
Raw: {"input_summary": "用户确认到场", "intention_id": "a1b2c3d4e5f6g7h8"}...


Evaluating:   5%|▍         | 23/506 [00:49<17:18,  2.15s/it]


--- Prediction 45 ---
Expected: c3d4e5f6a7b8c9d0
Predicted: c3d4e5f6a7b8c9d0
Raw: {"input_summary": "用户询问儿童门票", "intention_id": "c3d4e5f6a7b8c9d0"}...

--- Prediction 46 ---
Expected: z1x2c3v4b5n6m7q8
Predicted: z1x2c3v4b5n6m7q8
Raw: {"input_summary": "用户询问参展品牌", "intention_id": "z1x2c3v4b5n6m7q8"}...


Evaluating:   5%|▍         | 24/506 [00:52<17:15,  2.15s/it]


--- Prediction 47 ---
Expected: k_base_002
Predicted: k_base_002
Raw: {"input_summary": "用户询问活动地点", "intention_id": "k_base_002"}...

--- Prediction 48 ---
Expected: q7r8s9t0u1v2w3x4
Predicted: y5z6a7b8c9d0e1f2
Raw: {"input_summary": "用户暂时不需要", "intention_id": "y5z6a7b8c9d0e1f2"}...


Evaluating:   5%|▍         | 25/506 [00:54<17:25,  2.17s/it]


--- Prediction 49 ---
Expected: f9g0h1i2j3k4l5m6
Predicted: f9g0h1i2j3k4l5m6
Raw: {"input_summary": "用户询问以旧换新", "intention_id": "f9g0h1i2j3k4l5m6"}...

--- Prediction 50 ---
Expected: u4t3s2r1q0p9o8n7
Predicted: w6v5u4t3s2r1q0p9
Raw: {"input_summary": "用户询问展会具体信息", "intention_id": "w6v5u4t3s2r1q0p9"}...


Evaluating:   5%|▌         | 26/506 [00:56<17:29,  2.19s/it]


--- Prediction 51 ---
Expected: afea953927904771
Predicted: kb2349efcdb8917ec
Raw: {"input_summary": "用户抱怨电话骚扰", "intention_id": "kb2349efcdb8917ec"}...

--- Prediction 52 ---
Expected: others
Predicted: e5f67890a1b2c3d4
Raw: {"input_summary": "用户询问吃饭问题", "intention_id": "e5f67890a1b2c3d4"}...


Evaluating:   5%|▌         | 27/506 [00:58<17:16,  2.16s/it]


--- Prediction 53 ---
Expected: k5e6f7g8h9i0j1k2
Predicted: k5e6f7g8h9i0j1k2
Raw: {"input_summary": "用户孩子都有了", "intention_id": "k5e6f7g8h9i0j1k2"}...

--- Prediction 54 ---
Expected: k9l8m7n6o5p4q3r2
Predicted: k9l8m7n6o5p4q3r2
Raw: {"input_summary": "用户询问研讨会门票", "intention_id": "k9l8m7n6o5p4q3r2"}...


Evaluating:   6%|▌         | 28/506 [01:00<17:10,  2.16s/it]


--- Prediction 55 ---
Expected: z9y8x7w6v5u4t3s
Predicted: z9y8x7w6v5u4t3s
Raw: {"input_summary": "用户询问展馆位置", "intention_id": "z9y8x7w6v5u4t3s"}...

--- Prediction 56 ---
Expected: bace9d38206311ba
Predicted: bace9d38206311ba
Raw: {"input_summary": "用户挑逗客服", "intention_id": "bace9d38206311ba"}...


Evaluating:   6%|▌         | 29/506 [01:02<17:05,  2.15s/it]


--- Prediction 57 ---
Expected: 6543defg8765hijk
Predicted: 6543defg8765hijk
Raw: {"input_summary": "用户询问展出品牌", "intention_id": "6543defg8765hijk"}...

--- Prediction 58 ---
Expected: others
Predicted: 6a7b8c9d0e1f2345
Raw: {"input_summary": "用户询问酒店预定", "intention_id": "6a7b8c9d0e1f2345"}...


Evaluating:   6%|▌         | 30/506 [01:05<17:14,  2.17s/it]


--- Prediction 59 ---
Expected: 9c0v1b2n3m4q5w6e
Predicted: 2g3h4j5k6l7z8x9c
Raw: {"input_summary": "用户询问领取方式", "intention_id": "2g3h4j5k6l7z8x9c"}...

--- Prediction 60 ---
Expected: m3n4o5p6q7r8s9t0
Predicted: m3n4o5p6q7r8s9t0
Raw: {"input_summary": "用户询问服务是否免费", "intention_id": "m3n4o5p6q7r8s9t0"}...


Evaluating:   6%|▌         | 31/506 [01:07<17:07,  2.16s/it]


--- Prediction 61 ---
Expected: others
Predicted: o5p6q7r8s9t0u1v2
Raw: {"input_summary": "用户表达不满", "intention_id": "o5p6q7r8s9t0u1v2"}...

--- Prediction 62 ---
Expected: w2q3a4s5z6x7d8c9
Predicted: w2q3a4s5z6x7d8c9
Raw: {"input_summary": "用户询问停车情况", "intention_id": "w2q3a4s5z6x7d8c9"}...


Evaluating:   6%|▋         | 32/506 [01:09<17:02,  2.16s/it]


--- Prediction 63 ---
Expected: others
Predicted: k303
Raw: {"input_summary": "用户询问样板间", "intention_id": "k303"}...

--- Prediction 64 ---
Expected: l1k2l3m4n5o6p7q8
Predicted: l1k2l3m4n5o6p7q8
Raw: {"input_summary": "用户问抽奖活动", "intention_id": "l1k2l3m4n5o6p7q8"}...


Evaluating:   7%|▋         | 33/506 [01:11<17:03,  2.16s/it]


--- Prediction 65 ---
Expected: a1b2c3d4e5f6g7h8
Predicted: l1k2j3h4g5f6d7s8
Raw: {"input_summary": "用户表示放弃", "intention_id": "l1k2j3h4g5f6d7s8"}...

--- Prediction 66 ---
Expected: w1q2a3s4z5x6c7v8
Predicted: w1q2a3s4z5x6c7v8
Raw: {"input_summary": "用户询问展位费", "intention_id": "w1q2a3s4z5x6c7v8"}...


Evaluating:   7%|▋         | 34/506 [01:13<16:55,  2.15s/it]


--- Prediction 67 ---
Expected: kbn_018_abc123def
Predicted: kbn_018_abc123def
Raw: {"input_summary": "用户询问分期付款", "intention_id": "kbn_018_abc123def"}...

--- Prediction 68 ---
Expected: 2b3c4d5e6f7a8b9c
Predicted: 2b3c4d5e6f7a8b9c
Raw: {"input_summary": "用户询问展会时间", "intention_id": "2b3c4d5e6f7a8b9c"}...


Evaluating:   7%|▋         | 35/506 [01:15<16:45,  2.14s/it]


--- Prediction 69 ---
Expected: others
Predicted: others
Raw: {"input_summary": "用户言语不文明", "intention_id": "others"}...

--- Prediction 70 ---
Expected: f8a9b0c1d2e3f4a5
Predicted: f8a9b0c1d2e3f4a5
Raw: {"input_summary": "用户拒绝参加", "intention_id": "f8a9b0c1d2e3f4a5"}...


Evaluating:   7%|▋         | 36/506 [01:17<16:41,  2.13s/it]


--- Prediction 71 ---
Expected: g3h4i5j6k7l8m9n0
Predicted: g3h4i5j6k7l8m9n0
Raw: {"input_summary": "用户询问展会地点", "intention_id": "g3h4i5j6k7l8m9n0"}...

--- Prediction 72 ---
Expected: others
Predicted: others
Raw: {"input_summary": "用户询问带宠物", "intention_id": "others"}...


Evaluating:   7%|▋         | 37/506 [01:20<16:45,  2.14s/it]


--- Prediction 73 ---
Expected: z1y2x3w4v5u6t7s8
Predicted: others
Raw: {"input_summary": "用户约定见面", "intention_id": "others"}...

--- Prediction 74 ---
Expected: l2m3n4o5p6q7r8s9
Predicted: l2m3n4o5p6q7r8s9
Raw: {"input_summary": "用户调戏AI", "intention_id": "l2m3n4o5p6q7r8s9"}...


Evaluating:   8%|▊         | 38/506 [01:22<16:40,  2.14s/it]


--- Prediction 75 ---
Expected: c3b4a5c6b7a8c9b0
Predicted: c3b4a5c6b7a8c9b0
Raw: {"input_summary": "用户不想参加", "intention_id": "c3b4a5c6b7a8c9b0"}...

--- Prediction 76 ---
Expected: 9a8b7c6d5e4f3g2h1
Predicted: 9a8b7c6d5e4f3g2h1
Raw: {"input_summary": "用户想去看看", "intention_id": "9a8b7c6d5e4f3g2h1"}...


Evaluating:   8%|▊         | 39/506 [01:24<16:37,  2.14s/it]


--- Prediction 77 ---
Expected: s9t0u1v2w3x4y5z6
Predicted: s9t0u1v2w3x4y5z6
Raw: {"input_summary": "用户骂人", "intention_id": "s9t0u1v2w3x4y5z6"}...

--- Prediction 78 ---
Expected: others
Predicted: others
Raw: {"input_summary": "用户提及股市", "intention_id": "others"}...


Evaluating:   8%|▊         | 40/506 [01:26<16:47,  2.16s/it]


--- Prediction 79 ---
Expected: b2c3v4b5n6m7q8w9
Predicted: b2c3v4b5n6m7q8w9
Raw: {"input_summary": "用户对活动感兴趣", "intention_id": "b2c3v4b5n6m7q8w9"}...

--- Prediction 80 ---
Expected: 7p8q9r0s1t2u3v4w
Predicted: 7p8q9r0s1t2u3v4w
Raw: {"input_summary": "用户询问试驾", "intention_id": "7p8q9r0s1t2u3v4w"}...


Evaluating:   8%|▊         | 41/506 [01:28<15:55,  2.05s/it]


--- Prediction 81 ---
Expected: others
Predicted: 0abcdef123456789
Raw: {"input_summary": "用户询问信用卡办理", "intention_id": "0abcdef123456789"}...

--- Prediction 82 ---
Expected: k006
Predicted: k006
Raw: {"input_summary": "用户询问班车接送", "intention_id": "k006"}...


Evaluating:   8%|▊         | 42/506 [01:30<16:14,  2.10s/it]


--- Prediction 83 ---
Expected: z9x8c7v6b5n4m3q2
Predicted: z9x8c7v6b5n4m3q2
Raw: {"input_summary": "用户询问智能制造论坛", "intention_id": "z9x8c7v6b5n4m3q2"}...

--- Prediction 84 ---
Expected: k1l2m3n4o5p6q7r8s
Predicted: k1l2m3n4o5p6q7r8s
Raw: {"input_summary": "用户询问具体地点", "intention_id": "k1l2m3n4o5p6q7r8s"}...


Evaluating:   8%|▊         | 43/506 [01:32<16:14,  2.11s/it]


--- Prediction 85 ---
Expected: b8f9c2d1e4a3f5g6
Predicted: b8f9c2d1e4a3f5g6
Raw: {"input_summary": "用户态度不变", "intention_id": "b8f9c2d1e4a3f5g6"}...

--- Prediction 86 ---
Expected: z9y8x7w6v5u4t3s2
Predicted: z9y8x7w6v5u4t3s2
Raw: {"input_summary": "用户询问招商方式", "intention_id": "z9y8x7w6v5u4t3s2"}...


Evaluating:   9%|▊         | 44/506 [01:34<16:26,  2.13s/it]


--- Prediction 87 ---
Expected: K_5566778899001122
Predicted: K_5566778899001122
Raw: {"input_summary": "用户询问携带物品", "intention_id": "K_5566778899001122"}...

--- Prediction 88 ---
Expected: x9y8z7w6v5u4t3s2
Predicted: k006
Raw: {"input_summary": "用户想要礼品", "intention_id": "k006"}...


Evaluating:   9%|▉         | 45/506 [01:37<16:25,  2.14s/it]


--- Prediction 89 ---
Expected: kb_wedding_004
Predicted: kb_wedding_004
Raw: {"input_summary": "用户询问钻戒补贴", "intention_id": "kb_wedding_004"}...

--- Prediction 90 ---
Expected: 4d5e6f7g8h9i0j1k2
Predicted: 4d5e6f7g8h9i0j1k2
Raw: {"input_summary": "用户询问观众资格", "intention_id": "4d5e6f7g8h9i0j1k2"}...


Evaluating:   9%|▉         | 46/506 [01:39<16:08,  2.11s/it]


--- Prediction 91 ---
Expected: a1b2c3d4e5f6g7h8
Predicted: a1b2c3d4e5f6g7h8
Raw: {"input_summary": "用户拒绝参加", "intention_id": "a1b2c3d4e5f6g7h8"}...

--- Prediction 92 ---
Expected: k004
Predicted: k004
Raw: {"input_summary": "用户询问产品价格", "intention_id": "k004"}...


Evaluating:   9%|▉         | 47/506 [01:41<16:03,  2.10s/it]


--- Prediction 93 ---
Expected: a1b2c3d4e5f6a7b8
Predicted: a1b2c3d4e5f6a7b8
Raw: {"input_summary": "用户询问停车情况", "intention_id": "a1b2c3d4e5f6a7b8"}...

--- Prediction 94 ---
Expected: 8f3a2c1e9d4b5a67
Predicted: 8f3a2c1e9d4b5a67
Raw: {"input_summary": "用户确认参加", "intention_id": "8f3a2c1e9d4b5a67"}...


Evaluating:   9%|▉         | 48/506 [01:43<15:57,  2.09s/it]


--- Prediction 95 ---
Expected: a12b34c5d67e89f0
Predicted: f23a45b67c89d0e
Raw: {"input_summary": "用户有购车打算", "intention_id": "f23a45b67c89d0e"}...

--- Prediction 96 ---
Expected: others
Predicted: 64951aebf2e8c37d
Raw: {"input_summary": "用户拒绝客服", "intention_id": "64951aebf2e8c37d"}...


Evaluating:  10%|▉         | 49/506 [01:45<15:58,  2.10s/it]


--- Prediction 97 ---
Expected: others
Predicted: others
Raw: {"input_summary": "用户质疑活动合理性", "intention_id": "others"}...

--- Prediction 98 ---
Expected: 0j1k2l3m4n5o6p7q
Predicted: 0j1k2l3m4n5o6p7q
Raw: {"input_summary": "用户询问纪念品", "intention_id": "0j1k2l3m4n5o6p7q"}...


Evaluating:  10%|▉         | 50/506 [01:47<15:44,  2.07s/it]


--- Prediction 99 ---
Expected: k707070707070707
Predicted: k101010101010101
Raw: {"input_summary": "用户询问位置", "intention_id": "k101010101010101"}...

--- Prediction 100 ---
Expected: d4e5f6g7h8i9j0k1
Predicted: d4e5f6g7h8i9j0k1
Raw: {"input_summary": "用户询问门票", "intention_id": "d4e5f6g7h8i9j0k1"}...


Evaluating: 100%|██████████| 506/506 [18:08<00:00,  2.15s/it]


🔍 Skip reasons: {'no_messages': 0, 'json_parse': 0, 'no_id': 0, 'prompt_gen': 0, 'other': 0}

Base Qwen3-4B Results:
  Total processed: 1011
  Samples skipped: 0
  Correct: 776/1011 (76.76%)
  Valid responses: 100.00%


### 2.3 Save the base model performance

In [18]:
performance_base =  {
    "loss": base_loss,
    "perplexity": base_perplexity,
    "accuracy": base_results.get("accuracy", 0),
    "correct": base_results.get("correct", 0),
    "total": base_results.get("total", 0),
    "valid_rate": base_results.get("valid_rate", 0)
}

with open("performance_base.json", "w", encoding="utf-8") as f:
    json.dump(performance_base, f, ensure_ascii=False, indent=2)

## 3. Test the LoRA model

### 3.1 Load the LoRA adapter and configure the test

Load the best performance adapter during training: lr1e-4_3epoch

In [19]:
adapter_dir = "./lora_adapter_lr1e-4_3epoch"

In [20]:
print("\nLoading LoRA model...")
base_model = AutoModelForCausalLM.from_pretrained(
    MODEL_NAME,
    dtype=torch.bfloat16,
    device_map="cuda:0",
    trust_remote_code=True,
    attn_implementation="flash_attention_2"
)

lora_model = PeftModel.from_pretrained(
    base_model,
    adapter_dir,
    # ✅ Load adapter to CPU first to avoid GPU OOM
    torch_device="cpu",
)

# Now move the merged model to GPU
lora_model = lora_model.to("cuda:0")
lora_model.eval()

print(f"✅ LoRA model loaded from: {adapter_dir}")


Loading LoRA model...


Loading weights:   0%|          | 0/398 [00:00<?, ?it/s]

✅ LoRA model loaded from: ./lora_adapter


In [21]:
eval_args = TrainingArguments(
    output_dir="./temp_eval",
    per_device_eval_batch_size=1,  # ✅ Memory-safe
    dataloader_num_workers=0,
    bf16=True,
    report_to="none",
)

lora_trainer = Trainer(
    model=lora_model,
    args=eval_args,
    eval_dataset=tokenized_ds["test"],
    data_collator=data_collator
)

lora_trainer.remove_callback(NotebookProgressCallback)

### 3.2 Perform the test

In [22]:
print("\nComputing evaluation loss...")

lora_metrics = lora_trainer.evaluate()
lora_loss = lora_metrics["eval_loss"]
lora_perplexity = torch.exp(torch.tensor(lora_loss)).item()
print(f"LoRA Model - Loss: {lora_loss:.4f}, Perplexity: {lora_perplexity:.2f}")


Computing evaluation loss...
LoRA Model - Loss: 0.1198, Perplexity: 1.13


In [23]:
print("\nComputing generation accuracy...")

# ✅ Ensure left-padding for generation
original_padding = tokenizer.padding_side
tokenizer.padding_side = "left"

lora_results = evaluate_intention_accuracy(
    lora_model,
    tokenizer,
    ds["test"],  # Raw dataset with messages
    "LoRA Qwen3-4B",
    batch_size=2  # Memory-safe
)

# Restore padding
tokenizer.padding_side = original_padding


Computing generation accuracy...

Evaluating LoRA Qwen3-4B on intention_id...
  test_data type: <class 'datasets.arrow_dataset.Dataset'>
  test_data length: 1011


Evaluating:   0%|          | 1/506 [00:03<27:39,  3.29s/it]


--- Prediction 1 ---
Expected: 1a2b3c4d5e6f7g8h
Predicted: 1a2b3c4d5e6f7g8h
Raw: {"input_summary": "用户表示不在", "intention_id": "1a2b3c4d5e6f7g8h"}...

--- Prediction 2 ---
Expected: n0o1p2q3r4s5t6u7
Predicted: others
Raw: {"input_summary": "用户性骚扰客服", "intention_id": "others"}...


Evaluating:   0%|          | 2/506 [00:06<27:32,  3.28s/it]


--- Prediction 3 ---
Expected: 9z8y7x6w5v4u3t2s
Predicted: 9z8y7x6w5v4u3t2s
Raw: {"input_summary": "用户询问门票", "intention_id": "9z8y7x6w5v4u3t2s"}...

--- Prediction 4 ---
Expected: 1a2b3c4d5e6f7g8h
Predicted: others
Raw: {"input_summary": "询问能否带猫", "intention_id": "others"}...


Evaluating:   1%|          | 3/506 [00:10<28:42,  3.42s/it]


--- Prediction 5 ---
Expected: 1a2b3c4d5e6f7890
Predicted: 3c4d5e6f7890a1b2
Raw: {"input_summary": "用户表示忙抽不开身", "intention_id": "3c4d5e6f7890a1b2"}...

--- Prediction 6 ---
Expected: a1b2c3d4e5f6g7h8
Predicted: a1b2c3d4e5f6g7h8
Raw: {"input_summary": "用户拒绝参加", "intention_id": "a1b2c3d4e5f6g7h8"}...


Evaluating:   1%|          | 4/506 [00:13<28:40,  3.43s/it]


--- Prediction 7 ---
Expected: a1b2c3d4e5f6g7h8
Predicted: a1b2c3d4e5f6g7h8
Raw: {"input_summary": "用户厌烦拒绝", "intention_id": "a1b2c3d4e5f6g7h8"}...

--- Prediction 8 ---
Expected: h8i9j0k1l2m3n4o5
Predicted: h8i9j0k1l2m3n4o5
Raw: {"input_summary": "用户询问停车", "intention_id": "h8i9j0k1l2m3n4o5"}...


Evaluating:   1%|          | 5/506 [00:17<29:01,  3.48s/it]


--- Prediction 9 ---
Expected: d4e5f6g7h8i9j0k1l
Predicted: d4e5f6g7h8i9j0k1l
Raw: {"input_summary": "询问试驾是否收费", "intention_id": "d4e5f6g7h8i9j0k1l"}...

--- Prediction 10 ---
Expected: l2_n3_o4_p5_q6_r7
Predicted: l2_n3_o4_p5_q6_r7
Raw: {"input_summary": "询问是否收费", "intention_id": "l2_n3_o4_p5_q6_r7"}...


Evaluating:   1%|          | 6/506 [00:20<28:54,  3.47s/it]


--- Prediction 11 ---
Expected: d8e7f6a5b4c3d2e1
Predicted: d8e7f6a5b4c3d2e1
Raw: {"input_summary": "用户询问预约", "intention_id": "d8e7f6a5b4c3d2e1"}...

--- Prediction 12 ---
Expected: others
Predicted: i9j0k1l2m3n4o5p6
Raw: {"input_summary": "用户辱骂客服", "intention_id": "i9j0k1l2m3n4o5p6"}...


Evaluating:   1%|▏         | 7/506 [00:24<28:51,  3.47s/it]


--- Prediction 13 ---
Expected: y5z6a7b8c9d0e1f2
Predicted: y5z6a7b8c9d0e1f2
Raw: {"input_summary": "用户询问参展楼盘", "intention_id": "y5z6a7b8c9d0e1f2"}...

--- Prediction 14 ---
Expected: 9k0l1m2n3o4p5q6r
Predicted: 9k0l1m2n3o4p5q6r
Raw: {"input_summary": "用户肯定参加", "intention_id": "9k0l1m2n3o4p5q6r"}...


Evaluating:   2%|▏         | 8/506 [00:27<28:51,  3.48s/it]


--- Prediction 15 ---
Expected: others
Predicted: others
Raw: {"input_summary": "询问能否带宠物", "intention_id": "others"}...

--- Prediction 16 ---
Expected: c3d4e5f6g7h8i9j0
Predicted: c3d4e5f6g7h8i9j0
Raw: {"input_summary": "用户询问报名方式", "intention_id": "c3d4e5f6g7h8i9j0"}...


Evaluating:   2%|▏         | 9/506 [00:31<29:07,  3.52s/it]


--- Prediction 17 ---
Expected: 5e4f3a2b1c9d8e7f
Predicted: 5e4f3a2b1c9d8e7f
Raw: {"input_summary": "用户询问研讨会", "intention_id": "5e4f3a2b1c9d8e7f"}...

--- Prediction 18 ---
Expected: s9t0u1v2w3x4y5z6a
Predicted: s9t0u1v2w3x4y5z6a
Raw: {"input_summary": "用户询问媒体证", "intention_id": "s9t0u1v2w3x4y5z6a"}...


Evaluating:   2%|▏         | 10/506 [00:34<28:54,  3.50s/it]


--- Prediction 19 ---
Expected: b1c2d3e4f5g6h7i
Predicted: b1c2d3e4f5g6h7i
Raw: {"input_summary": "询问儿童区", "intention_id": "b1c2d3e4f5g6h7i"}...

--- Prediction 20 ---
Expected: others
Predicted: others
Raw: {"input_summary": "用户辱骂助手", "intention_id": "others"}...


Evaluating:   2%|▏         | 11/506 [00:38<29:11,  3.54s/it]


--- Prediction 21 ---
Expected: 5y6z7a8b9c0d1e2f
Predicted: 5y6z7a8b9c0d1e2f
Raw: {"input_summary": "用户询问交通", "intention_id": "5y6z7a8b9c0d1e2f"}...

--- Prediction 22 ---
Expected: k7007007007007007
Predicted: k7007007007007007
Raw: {"input_summary": "询问旧机回收", "intention_id": "k7007007007007007"}...


Evaluating:   2%|▏         | 12/506 [00:41<28:22,  3.45s/it]


--- Prediction 23 ---
Expected: kb_005_56789012
Predicted: kb_005_56789012
Raw: {"input_summary": "要求转人工", "intention_id": "kb_005_56789012"}...

--- Prediction 24 ---
Expected: others
Predicted: others
Raw: {"input_summary": "用户辱骂助手", "intention_id": "others"}...


Evaluating:   3%|▎         | 13/506 [00:44<28:16,  3.44s/it]


--- Prediction 25 ---
Expected: b2c3d4e5f6g7h8i9
Predicted: b2c3d4e5f6g7h8i9
Raw: {"input_summary": "询问开始时间", "intention_id": "b2c3d4e5f6g7h8i9"}...

--- Prediction 26 ---
Expected: d7e6f5g4h3i2j1k0
Predicted: d7e6f5g4h3i2j1k0
Raw: {"input_summary": "用户询问表演", "intention_id": "d7e6f5g4h3i2j1k0"}...


Evaluating:   3%|▎         | 14/506 [00:47<25:12,  3.07s/it]


--- Prediction 27 ---
Expected: others
Predicted: others
Raw: {"input_summary": "用户询问住宿", "intention_id": "others"}...

--- Prediction 28 ---
Expected: others
Predicted: others
Raw: {"input_summary": "用户询问是否管饭", "intention_id": "others"}...


Evaluating:   3%|▎         | 15/506 [00:50<26:15,  3.21s/it]


--- Prediction 29 ---
Expected: q1w2e3r4t5y6u7i8
Predicted: q1w2e3r4t5y6u7i8
Raw: {"input_summary": "询问另一个品牌", "intention_id": "q1w2e3r4t5y6u7i8"}...

--- Prediction 30 ---
Expected: k005
Predicted: 3c4d5e6f7890a1b2
Raw: {"input_summary": "用户索要资料", "intention_id": "3c4d5e6f7890a1b2"}...


Evaluating:   3%|▎         | 16/506 [00:54<27:42,  3.39s/it]


--- Prediction 31 ---
Expected: b3d8e1f4a6c9d7b2
Predicted: b3d8e1f4a6c9d7b2
Raw: {"input_summary": "用户需要时间考虑", "intention_id": "b3d8e1f4a6c9d7b2"}...

--- Prediction 32 ---
Expected: 9f8d7s6a5s4d3a2s
Predicted: 9f8d7s6a5s4d3a2s
Raw: {"input_summary": "用户拒绝并要求别再打扰", "intention_id": "9f8d7s6a5s4d3a2s"}...


Evaluating:   3%|▎         | 17/506 [00:58<28:05,  3.45s/it]


--- Prediction 33 ---
Expected: r8s7t6u5v4w3x2y1
Predicted: r8s7t6u5v4w3x2y1
Raw: {"input_summary": "用户询问门票", "intention_id": "r8s7t6u5v4w3x2y1"}...

--- Prediction 34 ---
Expected: others
Predicted: others
Raw: {"input_summary": "用户性骚扰客服", "intention_id": "others"}...


Evaluating:   4%|▎         | 18/506 [01:01<28:11,  3.47s/it]


--- Prediction 35 ---
Expected: others
Predicted: others
Raw: {"input_summary": "用户询问住宿", "intention_id": "others"}...

--- Prediction 36 ---
Expected: m2n3b4v5c6x7z8l9k0
Predicted: m2n3b4v5c6x7z8l9k0
Raw: {"input_summary": "用户询问收费", "intention_id": "m2n3b4v5c6x7z8l9k0"}...


Evaluating:   4%|▍         | 19/506 [01:05<28:47,  3.55s/it]


--- Prediction 37 ---
Expected: a3b4c5d6e7f8a9b0
Predicted: a3b4c5d6e7f8a9b0
Raw: {"input_summary": "询问旧物回收", "intention_id": "a3b4c5d6e7f8a9b0"}...

--- Prediction 38 ---
Expected: 71e25d86c5f68440
Predicted: 71e25d86c5f68440
Raw: {"input_summary": "用户辱骂客服", "intention_id": "71e25d86c5f68440"}...


Evaluating:   4%|▍         | 20/506 [01:08<28:21,  3.50s/it]


--- Prediction 39 ---
Expected: 6f7g8h9i0j1k2l3m
Predicted: 6f7g8h9i0j1k2l3m
Raw: {"input_summary": "用户辱骂客服", "intention_id": "6f7g8h9i0j1k2l3m"}...

--- Prediction 40 ---
Expected: others
Predicted: others
Raw: {"input_summary": "用户询问证件", "intention_id": "others"}...


Evaluating:   4%|▍         | 21/506 [01:12<28:00,  3.46s/it]


--- Prediction 41 ---
Expected: 3a2b1c9d8e7f6g5h
Predicted: 3a2b1c9d8e7f6g5h
Raw: {"input_summary": "用户询问买样品", "intention_id": "3a2b1c9d8e7f6g5h"}...

--- Prediction 42 ---
Expected: k_a1b2c3d4e5f6g7h
Predicted: k_a1b2c3d4e5f6g7h
Raw: {"input_summary": "询问门票费用", "intention_id": "k_a1b2c3d4e5f6g7h"}...


Evaluating:   4%|▍         | 22/506 [01:15<28:11,  3.49s/it]


--- Prediction 43 ---
Expected: h2j3k4l5m6n7b8v9c
Predicted: h2j3k4l5m6n7b8v9c
Raw: {"input_summary": "用户询问班次", "intention_id": "h2j3k4l5m6n7b8v9c"}...

--- Prediction 44 ---
Expected: a1b2c3d4e5f6g7h8
Predicted: a1b2c3d4e5f6g7h8
Raw: {"input_summary": "用户确认到场", "intention_id": "a1b2c3d4e5f6g7h8"}...


Evaluating:   5%|▍         | 23/506 [01:19<27:50,  3.46s/it]


--- Prediction 45 ---
Expected: c3d4e5f6a7b8c9d0
Predicted: c3d4e5f6a7b8c9d0
Raw: {"input_summary": "询问儿童门票", "intention_id": "c3d4e5f6a7b8c9d0"}...

--- Prediction 46 ---
Expected: z1x2c3v4b5n6m7q8
Predicted: z1x2c3v4b5n6m7q8
Raw: {"input_summary": "用户询问参展品牌", "intention_id": "z1x2c3v4b5n6m7q8"}...


Evaluating:   5%|▍         | 24/506 [01:22<27:33,  3.43s/it]


--- Prediction 47 ---
Expected: k_base_002
Predicted: k_base_002
Raw: {"input_summary": "询问活动地点", "intention_id": "k_base_002"}...

--- Prediction 48 ---
Expected: q7r8s9t0u1v2w3x4
Predicted: q7r8s9t0u1v2w3x4
Raw: {"input_summary": "用户表示不需要", "intention_id": "q7r8s9t0u1v2w3x4"}...


Evaluating:   5%|▍         | 25/506 [01:25<27:52,  3.48s/it]


--- Prediction 49 ---
Expected: f9g0h1i2j3k4l5m6
Predicted: f9g0h1i2j3k4l5m6
Raw: {"input_summary": "用户询问以旧换新", "intention_id": "f9g0h1i2j3k4l5m6"}...

--- Prediction 50 ---
Expected: u4t3s2r1q0p9o8n7
Predicted: u4t3s2r1q0p9o8
Raw: {"input_summary": "询问报名方式", "intention_id": "u4t3s2r1q0p9o8"}...


Evaluating:   5%|▌         | 26/506 [01:29<26:57,  3.37s/it]


--- Prediction 51 ---
Expected: afea953927904771
Predicted: kb761cdc4ea173f04
Raw: {"input_summary": "用户辱骂客服", "intention_id": "kb761cdc4ea173f04"}...

--- Prediction 52 ---
Expected: others
Predicted: others
Raw: {"input_summary": "用户询问是否管饭", "intention_id": "others"}...


Evaluating:   5%|▌         | 27/506 [01:32<27:21,  3.43s/it]


--- Prediction 53 ---
Expected: k5e6f7g8h9i0j1k2
Predicted: k5e6f7g8h9i0j1k2
Raw: {"input_summary": "用户抱怨孩子已大", "intention_id": "k5e6f7g8h9i0j1k2"}...

--- Prediction 54 ---
Expected: k9l8m7n6o5p4q3r2
Predicted: k9l8m7n6o5p4q3r2
Raw: {"input_summary": "询问研讨会费用", "intention_id": "k9l8m7n6o5p4q3r2"}...


Evaluating:   6%|▌         | 28/506 [01:35<26:52,  3.37s/it]


--- Prediction 55 ---
Expected: z9y8x7w6v5u4t3s
Predicted: z9y8x7w6v5u4t3s
Raw: {"input_summary": "用户询问地点", "intention_id": "z9y8x7w6v5u4t3s"}...

--- Prediction 56 ---
Expected: bace9d38206311ba
Predicted: bace9d38206311ba
Raw: {"input_summary": "用户调戏客服", "intention_id": "bace9d38206311ba"}...


Evaluating:   6%|▌         | 29/506 [01:38<25:56,  3.26s/it]


--- Prediction 57 ---
Expected: 6543defg8765hijk
Predicted: 6543defg8765hijk
Raw: {"input_summary": "用户询问参展品牌", "intention_id": "6543defg8765hijk"}...

--- Prediction 58 ---
Expected: others
Predicted: others
Raw: {"input_summary": "询问酒店预订", "intention_id": "others"}...


Evaluating:   6%|▌         | 30/506 [01:42<26:19,  3.32s/it]


--- Prediction 59 ---
Expected: 9c0v1b2n3m4q5w6e
Predicted: 9c0v1b2n3m4q5w6e
Raw: {"input_summary": "用户询问领取方式", "intention_id": "9c0v1b2n3m4q5w6e"}...

--- Prediction 60 ---
Expected: m3n4o5p6q7r8s9t0
Predicted: m3n4o5p6q7r8s9t0
Raw: {"input_summary": "询问服务是否收费", "intention_id": "m3n4o5p6q7r8s9t0"}...


Evaluating:   6%|▌         | 31/506 [01:45<26:13,  3.31s/it]


--- Prediction 61 ---
Expected: others
Predicted: others
Raw: {"input_summary": "用户辱骂客服", "intention_id": "others"}...

--- Prediction 62 ---
Expected: w2q3a4s5z6x7d8c9
Predicted: w2q3a4s5z6x7d8c9
Raw: {"input_summary": "用户询问停车", "intention_id": "w2q3a4s5z6x7d8c9"}...


Evaluating:   6%|▋         | 32/506 [01:49<26:21,  3.34s/it]


--- Prediction 63 ---
Expected: others
Predicted: others
Raw: {"input_summary": "询问样板间", "intention_id": "others"}...

--- Prediction 64 ---
Expected: l1k2l3m4n5o6p7q8
Predicted: l1k2l3m4n5o6p7q8
Raw: {"input_summary": "询问抽奖活动", "intention_id": "l1k2l3m4n5o6p7q8"}...


Evaluating:   7%|▋         | 33/506 [01:52<26:35,  3.37s/it]


--- Prediction 65 ---
Expected: a1b2c3d4e5f6g7h8
Predicted: a1b2c3d4e5f6g7h8
Raw: {"input_summary": "用户同意参加", "intention_id": "a1b2c3d4e5f6g7h8"}...

--- Prediction 66 ---
Expected: w1q2a3s4z5x6c7v8
Predicted: w1q2a3s4z5x6c7v8
Raw: {"input_summary": "用户询问展位费", "intention_id": "w1q2a3s4z5x6c7v8"}...


Evaluating:   7%|▋         | 34/506 [01:55<26:18,  3.34s/it]


--- Prediction 67 ---
Expected: kbn_018_abc123def
Predicted: kbn_018_abc123def
Raw: {"input_summary": "询问分期付款", "intention_id": "kbn_018_abc123def"}...

--- Prediction 68 ---
Expected: 2b3c4d5e6f7a8b9c
Predicted: 2b3c4d5e6f7a8b9c
Raw: {"input_summary": "用户询问时间", "intention_id": "2b3c4d5e6f7a8b9c"}...


Evaluating:   7%|▋         | 35/506 [01:59<26:16,  3.35s/it]


--- Prediction 69 ---
Expected: others
Predicted: others
Raw: {"input_summary": "用户性骚扰调戏", "intention_id": "others"}...

--- Prediction 70 ---
Expected: f8a9b0c1d2e3f4a5
Predicted: f8a9b0c1d2e3f4a5
Raw: {"input_summary": "用户表示忙", "intention_id": "f8a9b0c1d2e3f4a5"}...


Evaluating:   7%|▋         | 36/506 [02:02<26:27,  3.38s/it]


--- Prediction 71 ---
Expected: g3h4i5j6k7l8m9n0
Predicted: g3h4i5j6k7l8m9n0
Raw: {"input_summary": "用户询问地点", "intention_id": "g3h4i5j6k7l8m9n0"}...

--- Prediction 72 ---
Expected: others
Predicted: others
Raw: {"input_summary": "用户询问能否带宠物", "intention_id": "others"}...


Evaluating:   7%|▋         | 37/506 [02:06<26:42,  3.42s/it]


--- Prediction 73 ---
Expected: z1y2x3w4v5u6t7s8
Predicted: z1y2x3w4v5u6t7s8
Raw: {"input_summary": "用户有意向参加", "intention_id": "z1y2x3w4v5u6t7s8"}...

--- Prediction 74 ---
Expected: l2m3n4o5p6q7r8s9
Predicted: l2m3n4o5p6q7r8s9
Raw: {"input_summary": "用户调戏客服", "intention_id": "l2m3n4o5p6q7r8s9"}...


Evaluating:   8%|▊         | 38/506 [02:09<27:04,  3.47s/it]


--- Prediction 75 ---
Expected: c3b4a5c6b7a8c9b0
Predicted: c3b4a5c6b7a8c9b0
Raw: {"input_summary": "用户赶不上", "intention_id": "c3b4a5c6b7a8c9b0"}...

--- Prediction 76 ---
Expected: 9a8b7c6d5e4f3g2h1
Predicted: 9a8b7c6d5e4f3g2h1
Raw: {"input_summary": "用户有意向参加", "intention_id": "9a8b7c6d5e4f3g2h1"}...


Evaluating:   8%|▊         | 39/506 [02:11<23:57,  3.08s/it]


--- Prediction 77 ---
Expected: s9t0u1v2w3x4y5z6
Predicted: others
Raw: {"input_summary": "用户辱骂客服", "intention_id": "others"}...

--- Prediction 78 ---
Expected: others
Predicted: others
Raw: {"input_summary": "用户抱怨股市", "intention_id": "others"}...


Evaluating:   8%|▊         | 40/506 [02:15<25:03,  3.23s/it]


--- Prediction 79 ---
Expected: b2c3v4b5n6m7q8w9
Predicted: b2c3v4b5n6m7q8w9
Raw: {"input_summary": "用户对讲座感兴趣", "intention_id": "b2c3v4b5n6m7q8w9"}...

--- Prediction 80 ---
Expected: 7p8q9r0s1t2u3v4w
Predicted: 7p8q9r0s1t2u3v4w
Raw: {"input_summary": "用户询问试驾", "intention_id": "7p8q9r0s1t2u3v4w"}...


Evaluating:   8%|▊         | 41/506 [02:17<22:41,  2.93s/it]


--- Prediction 81 ---
Expected: others
Predicted: others
Raw: {"input_summary": "询问是否支持信用卡", "intention_id": "others"}...

--- Prediction 82 ---
Expected: k006
Predicted: k006
Raw: {"input_summary": "询问是否有班车", "intention_id": "k006"}...


Evaluating:   8%|▊         | 42/506 [02:21<24:01,  3.11s/it]


--- Prediction 83 ---
Expected: z9x8c7v6b5n4m3q2
Predicted: z9x8c7v6b5n4m3q2
Raw: {"input_summary": "用户询问论坛", "intention_id": "z9x8c7v6b5n4m3q2"}...

--- Prediction 84 ---
Expected: k1l2m3n4o5p6q7r8s
Predicted: k1l2m3n4o5p6q7r8s
Raw: {"input_summary": "用户询问地点", "intention_id": "k1l2m3n4o5p6q7r8s"}...


Evaluating:   8%|▊         | 43/506 [02:24<25:01,  3.24s/it]


--- Prediction 85 ---
Expected: b8f9c2d1e4a3f5g6
Predicted: b8f9c2d1e4a3f5g6
Raw: {"input_summary": "用户坚持没时间", "intention_id": "b8f9c2d1e4a3f5g6"}...

--- Prediction 86 ---
Expected: z9y8x7w6v5u4t3s2
Predicted: z9y8x7w6v5u4t3s2
Raw: {"input_summary": "询问招商参展", "intention_id": "z9y8x7w6v5u4t3s2"}...


Evaluating:   9%|▊         | 44/506 [02:28<26:10,  3.40s/it]


--- Prediction 87 ---
Expected: K_5566778899001122
Predicted: K_5566778899001122
Raw: {"input_summary": "询问是否带U盘", "intention_id": "K_5566778899001122"}...

--- Prediction 88 ---
Expected: x9y8z7w6v5u4t3s2
Predicted: k006
Raw: {"input_summary": "用户想要领礼品", "intention_id": "k006"}...


Evaluating:   9%|▉         | 45/506 [02:31<25:57,  3.38s/it]


--- Prediction 89 ---
Expected: kb_wedding_004
Predicted: kb_wedding_004
Raw: {"input_summary": "询问钻戒补贴", "intention_id": "kb_wedding_004"}...

--- Prediction 90 ---
Expected: 4d5e6f7g8h9i0j1k2
Predicted: 4d5e6f7g8h9i0j1k2
Raw: {"input_summary": "询问观众资格", "intention_id": "4d5e6f7g8h9i0j1k2"}...


Evaluating:   9%|▉         | 46/506 [02:35<25:43,  3.36s/it]


--- Prediction 91 ---
Expected: a1b2c3d4e5f6g7h8
Predicted: a1b2c3d4e5f6g7h8
Raw: {"input_summary": "用户拒绝骚扰", "intention_id": "a1b2c3d4e5f6g7h8"}...

--- Prediction 92 ---
Expected: k004
Predicted: k004
Raw: {"input_summary": "询问云朵沙发价格", "intention_id": "k004"}...


Evaluating:   9%|▉         | 47/506 [02:38<25:42,  3.36s/it]


--- Prediction 93 ---
Expected: a1b2c3d4e5f6a7b8
Predicted: a1b2c3d4e5f6a7b8
Raw: {"input_summary": "询问停车情况", "intention_id": "a1b2c3d4e5f6a7b8"}...

--- Prediction 94 ---
Expected: 8f3a2c1e9d4b5a67
Predicted: 8f3a2c1e9d4b5a67
Raw: {"input_summary": "用户同意参加", "intention_id": "8f3a2c1e9d4b5a67"}...


Evaluating:   9%|▉         | 48/506 [02:41<25:47,  3.38s/it]


--- Prediction 95 ---
Expected: a12b34c5d67e89f0
Predicted: a12b34c5d67e89f0
Raw: {"input_summary": "用户有购车打算", "intention_id": "a12b34c5d67e89f0"}...

--- Prediction 96 ---
Expected: others
Predicted: 12bc7fa9de063514
Raw: {"input_summary": "用户辱骂客服", "intention_id": "12bc7fa9de063514"}...


Evaluating:  10%|▉         | 49/506 [02:45<25:39,  3.37s/it]


--- Prediction 97 ---
Expected: others
Predicted: others
Raw: {"input_summary": "用户辱骂助手", "intention_id": "others"}...

--- Prediction 98 ---
Expected: 0j1k2l3m4n5o6p7q
Predicted: 0j1k2l3m4n5o6p7q
Raw: {"input_summary": "用户询问纪念品", "intention_id": "0j1k2l3m4n5o6p7q"}...


Evaluating:  10%|▉         | 50/506 [02:48<25:18,  3.33s/it]


--- Prediction 99 ---
Expected: k707070707070707
Predicted: k707070707070707
Raw: {"input_summary": "询问老场馆", "intention_id": "k707070707070707"}...

--- Prediction 100 ---
Expected: d4e5f6g7h8i9j0k1
Predicted: d4e5f6g7h8i9j0k1
Raw: {"input_summary": "用户询问门票", "intention_id": "d4e5f6g7h8i9j0k1"}...


Evaluating: 100%|██████████| 506/506 [26:58<00:00,  3.20s/it]


🔍 Skip reasons: {'no_messages': 0, 'json_parse': 0, 'no_id': 0, 'prompt_gen': 0, 'other': 0}

LoRA Qwen3-4B Results:
  Total processed: 1011
  Samples skipped: 0
  Correct: 941/1011 (93.08%)
  Valid responses: 100.00%


### 3.3 Save the LoRA model performance

In [24]:
performance_lora =  {
    "loss": lora_loss,
    "perplexity": lora_perplexity,
    "accuracy": lora_results.get("accuracy", 0),
    "correct": lora_results.get("correct", 0),
    "total": lora_results.get("total", 0),
    "valid_rate": lora_results.get("valid_rate", 0)
}

with open("performance_lora.json", "w", encoding="utf-8") as f:
    json.dump(performance_lora, f, ensure_ascii=False, indent=2)

## 4. Final comparison

In [ ]:
with open("performance_base.json", "r", encoding="utf-8") as f:
    performance_base = json.load(f)

In [25]:
performance_comparison = {
    "base_model": performance_base,
    "lora_model": performance_lora,
    "improvement": {
        "loss_reduction": (performance_base["loss"]-performance_lora["loss"])/performance_base["loss"],
        "perplexity_reduction": (performance_base["perplexity"]-performance_lora["perplexity"])/performance_base["perplexity"],
        "accuracy_gain": performance_lora['accuracy']-performance_base['accuracy'],
        "valid_rate_gain": performance_lora['valid_rate']-performance_base['valid_rate']
    }
}

In [26]:
with open("performance_comparison.json", "w", encoding='utf-8') as f:
    json.dump(performance_comparison, f, indent=2, ensure_ascii=False)

print(f"\n✅ Comparison results saved to: {"performance_comparison.json"}")


✅ Comparison results saved to: performance_comparison.json


In [27]:
# ----------------------------
# Compare Results
# ----------------------------
print("\n" + "=" * 60)
print("FINAL COMPARISON: Base vs LoRA")
print("=" * 60)
print(f"{'Metric':<25} {'Base Model':<15} {'LoRA Model':<15} {'Improvement':<15}")
print("-" * 60)
print(f"{'Eval Loss':<25} {performance_base["loss"]:.4f}{'':<10} {performance_lora["loss"]:.4f}{'':<10} {(performance_base["loss"]-performance_lora["loss"])/performance_base["loss"]*100:+.2f}%")
print(f"{'Perplexity':<25} {performance_base["perplexity"]:.2f}{'':<10} {performance_lora["perplexity"]:.2f}{'':<10} {(performance_base["perplexity"]-performance_lora["perplexity"])/performance_base["perplexity"]*100:+.2f}%")
print(f"{'Accuracy':<25} {performance_base["accuracy"]:.2f}%{'':<7} {performance_lora['accuracy']:.2f}%{'':<7} {performance_lora['accuracy']-performance_base['accuracy']:+.2f}%")
print(f"{'Valid Rate':<25} {performance_base["valid_rate"]:.2f}%{'':<7} {performance_lora['valid_rate']:.2f}%{'':<7} {performance_lora['valid_rate']-performance_base['valid_rate']:+.2f}%")
print("=" * 60)


FINAL COMPARISON: Base vs LoRA
Metric                    Base Model      LoRA Model      Improvement    
------------------------------------------------------------
Eval Loss                 1.8264           0.1198           +93.44%
Perplexity                6.21           1.13           +81.85%
Accuracy                  76.76%        93.08%        +16.32%
Valid Rate                100.00%        100.00%        +0.00%


**As show in the table, the LoRA model increased the accuracy from fairly risky 76.76% to the production grade of 93.08%.**  
This is a clear proof of the success of the fine-tuning to make the model more specialized for intention identification.